docker/jupyter/custom/JupyterLockingExtension.js

```diff
@@ -0,0 +1,325 @@
+        }
+
+        renderModeBanner(isEditMode)
+    }
+
+    async function getLock() {
+        const payload = Object.assign(basePayload, {
+            method: 'POST',
+            body: JSON.stringify(getLocalPath())
+        })
+
+        fetch(lockUrl, payload)
+            .then(res => {
+                handleLockStatus(res)
+                return res.json()
+            })
+            .catch(err => {
+                console.log('in catch for getLock')
+                console.error(err)
+            })
+    }
+
+    async function toggleMetaFailureBanner(shouldShow) {
+        const bannerId = "notification_metaFailure"
+        const bannerText = "Failed to check notebook status, changes may not be saved to workspace. Retrying..."
+
+        if ($("#" + bannerId).length > 0) {
+            $("#" + bannerId).remove() //always remove because if we show we always re-render
+        }
+
+        if (shouldShow) {
+            const bannerStyling = "btn btn-xs navbar-btn btn-danger"
+
+            $('#notification_area').append(
+                $('<div>').attr({
+                    "id": bannerId,
+                    "class": bannerStyling
+                })
+                .append($('<span>').html('<i class="fa fa-exclamation-triangle"></i> ' + bannerText))
+            );
+        }
+
+    }
+
+    async function handleLockStatus(res) {
+        if (!res.ok) {
+            const status = res.status
+            const errorText = res.statusText
+
+            if (status == 409) {
+                res.json().then(body => {
+                    const message = getLockConflictBody(body.lockedBy)
+                    promptUserWithModal(lockConflictTitle, lockIssueButtons, message)
+                })
+            }
+            //for the lock endpoint, we consider all non 'ok' statuses an error
+            throw new Error(errorText)
+        }
+    }
+
+    const getLockConflictBody = (lockHolder) => `<p>This file is currently being editted by ${lockHolder}.</p><br/><p>You can make a copy, or run it in Playground Mode to explore and execute its contents without saving any changes.`
+
+    async function promptUserWithModal(title, buttons, htmlBody) {
+        if (modalOpen) return
+
+        modalOpen = true
+
+        dialog.modal({
+                body: $('<p>').html(htmlBody),
+                title: title,
+                buttons: buttons,
+                notebook: Jupyter.notebook,
+                keyboard_manager: Jupyter.notebook.keyboard_manager
+            })
+            .on('hidden.bs.modal', () => modalOpen = false)
+            .attr('id', 'leoUserModal')
+            .find(".close").click(() => window.history.back()) //TODO: test in docker image
+    }
+
+    async function openPlaygroundMode() {
+        const url = jupyterContentsAPIUrl + Jupyter.notebook.notebook_path
+        const newPath = meta.storageLink.localSafeModeBaseDirectory.replace('notebooks/', '') + '/' + Jupyter.notebook.notebook_name
```

Comment:
Why do we need to remove `notebooks/` here?

Reply:
The jupyter contents API requires the local path to the file, but it assumes all notebooks are within notebooks/ and wants it omitted